tools/audit-web-app.js

```diff
@@ -0,0 +1,200 @@
+ *
+ * Usage:
+ * ```sh
+ * node scripts/audit-web-app <url> <min-scores> [<log-file>]
+ * ```
+ *
+ * Runs audits against the specified URL on specific categories (accessibility,
+ * best practices, performance, PWA, SEO). It fails, if the score in any
+ * category is below the score specified in `<min-scores>`. (Only runs audits
+ * for the specified categories.)
+ *
+ * `<min-scores>` is either a number (in which case it is interpreted as
+ * `all:<min-score>`) or a list of comma-separated strings of the form
+ * `key:value`, where `key` is one of `accessibility`, `best-practices`,
+ * `performance`, `pwa`, `seo` or `all` and `value` is a number (between 0 and
+ * 100).
+ *
+ * Examples:
+ * - `95` _(Same as `all:95`.)_
+ * - `all:95` _(Run audits for all categories and require a score of 95 or higher.)_
+ * - `all:95,pwa:100` _(Same as `all:95`, except that a score of 100 is required
+ *    for the `pwa` category.)_
+ * - `performance:90` _(Only run audits for the `performance` category and
+ *    require a score of 90 or higher.)_
+ *
+ * If `<log-file>` is defined, the full results will be logged there.
+ *
+ * (Skips HTTPS-related audits, when run for an HTTP URL.)
+ */
+
+// Imports
+const lighthouse = require('lighthouse');
+const printer = require('lighthouse/lighthouse-cli/printer');
+const logger = require('lighthouse-logger');
+const puppeteer = require('puppeteer');
+
+// Constants
+const AUDIT_CATEGORIES =
+    [ 'accessibility', 'best-practices', 'performance', 'pwa', 'seo' ];
+const LIGHTHOUSE_FLAGS = {
+  logLevel : process.env.CI ? 'error' : 'info'
+}; // Be less verbose on CI.
+const SKIPPED_HTTPS_AUDITS = [ 'redirects-http', 'uses-http2' ];
+const VIEWER_URL = 'https://googlechrome.github.io/lighthouse/viewer';
+const WAIT_FOR_SW_DELAY = 5000;
+
+// Run
+_main(process.argv.slice(2)).then(() => console.log('Audit completed.'));
+
+// Functions - Definitions
+async function _main(args) {
+  const {url, minScores, logFile} = parseInput(args);
+  const isOnHttp = /^http:/.test(url);
+  const lhFlags = {
+    ...LIGHTHOUSE_FLAGS,
+    onlyCategories : Object.keys(minScores).sort()
+  };
+  const lhConfig = {
+    extends : 'lighthouse:default',
+    // Since the Angular ServiceWorker waits for the app to stabilize before
+    // registering, wait a few seconds after load to allow Lighthouse to
+    // reliably detect it.
+    passes :
+        [ {passName : 'defaultPass', pauseAfterLoadMs : WAIT_FOR_SW_DELAY} ],
+  };
+
+  console.log(`Running web-app audits for '${url}'...`);
+  console.log(`  Audit categories: ${lhFlags.onlyCategories.join(', ')}`);
+
+  // If testing on HTTP, skip HTTPS-specific tests.
+  // (Note: Browsers special-case localhost and run ServiceWorker even on HTTP.)
+  if (isOnHttp)
+    skipHttpsAudits(lhConfig);
+
+  logger.setLevel(lhFlags.logLevel);
+
+  try {
+    console.log('');
```

Comment:
Is this necessary?  At minimum if we want a blank line, we can just have `console.log()`